upstairs/src/client.rs

```diff
@@ -0,0 +1,3133 @@
+        &mut self,
+        repair_id: ReconciliationId,
+        job: &mut ReconcileIO,
+    ) -> bool {
+        let old_state = job.state.insert(self.client_id, IOState::Done);
+        assert_eq!(old_state, IOState::InProgress);
+        assert_eq!(job.id, repair_id);
+        job.state
+            .iter()
+            .all(|s| matches!(s, IOState::Done | IOState::Skipped))
+    }
+
+    pub(crate) fn total_live_work(&self) -> usize {
+        (self.io_state_count.new + self.io_state_count.in_progress) as usize
+    }
+}
+
+/// How to handle "promote to active" requests
+#[derive(Debug)]
+enum PromoteState {
+    /// Send `PromoteToActive` when the state machine reaches `WaitForPromote`
+    Waiting(u64),
+    /// We have already sent `PromoteToActive`
+    Sent(u64),
+}
+
+/// Tracks client negotiation progress
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum NegotiationState {
+    Start,
+    WaitForPromote,
+    WaitForRegionInfo,
+    GetLastFlush,
+    GetExtentVersions,
+    Done,
+}
+
+/// Action requested by `DownstairsClient::select`
+///
+/// This is split into a separate data structure because we need to distinguish
+/// between choosing an action (which must be cancel safe) and applying that
+/// action (which need not be).
+#[derive(Debug)]
+pub(crate) enum ClientAction {
+    /// We have connected to the socket
+    Connected,
+
+    /// We have received a message on the client task channel
+    Response(Message),
+
+    /// The client task has stopped
+    TaskStopped(ClientRunResult),
+
+    /// It's time to ping the client
+    Ping,
+
+    /// The client has hit a (Crucible) timeout
+    Timeout,
+
+    /// The client IO channel has returned `None`
+    ///
+    /// This should never happen during normal operation, because
+    /// 1) the IO task continues to run until the main upstairs task tells it to
+    ///    stop (by dropping the `client_request_tx` handle), and
+    /// 2) we are running with `panic=abort`, so if the IO task panics, it
+    ///    should bring down the whole program
+    ///
+    /// However, this _may_ happen during shutdown, because the Tokio runtime
+    /// shuts down tasks in arbitrary order and it's possible for the IO channel
+    /// sender to be dropped before the main task stops.
+    ChannelClosed,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct DownstairsStats {
+    /// Number of errors recorded
+    pub downstairs_errors: usize,
+
+    /// Count of extents repaired live.
+    pub extents_repaired: usize,
+
+    /// Count of extents checked but not needing live repair.
+    pub extents_confirmed: usize,
+
+    /// Count of time a downstairs LiveRepair completed.
+    pub live_repair_completed: usize,
+
+    /// Count of time a downstairs LiveRepair was aborted.
+    pub live_repair_aborted: usize,
+
+    /// Times we skipped repairing a downstairs due to being read-only
+    pub ro_lr_skipped: usize,
+
+    /// Count of downstairs connections
```

Comment:
Do we count connected and/or replaced?  I could not find the `+= 1`

Reply:
- `stats.replaced` is incremented [here](https://github.com/oxidecomputer/crucible/blob/9c95e2c68868e34f32c2b21bb10b359b8d3cfcff/upstairs/src/client.rs#L855)
- `stats.connected` isn't incremented anywhere, good catch!  Fixed in 0f261a5e8b1628ba8e8563272dff832fc6cb9751